File: app/scrapers/netherlands.ts
import Bluebird from 'bluebird';
import fetch from 'node-fetch';

import {
  NETHERLANDS_MEASUREMENTS_URL,
  NETHERLANDS_STATIONS_URL,
} from '../constants/urls';
import { City } from '../db/models';
import { getLatLng } from '../helpers/geocoder';

interface Station {
  number: string;
  location: string;
}

interface Measurement {
  station_number: string;
  value: number;
  timestamp_measured: string;
  formula: string;
}

interface Response<TData> {
  pagination: {
    last_page: number;
    first_page: number;
    prev_page: number;
    current_page: number;
    page_list: string[];
    next_page: number;
  };
  data: TData;
}

export const scrape = async (): Promise<City[]> => {
  const stations: Station[] = [];

  // Fetch all stations
  let stationsPage = 1;

  while (true) {
    const result = await fetch(
      `${NETHERLANDS_STATIONS_URL}?page=${stationsPage}`
    );
    const json: Response<Station[]> = await result.json();

    stations.push(...json.data);
    stationsPage = json.pagination.next_page;
    if (
      json.pagination.current_page === json.pagination.last_page ||
      json.data.length === 0
    ) {
      break;
    }
  }

  console.log('Netherlands: Retrieved', stations.length, 'stations');

  const measurements: Measurement[] = [];
  let measurementsPage = 1;

  const currentTimeEpoch = new Date().valueOf();
  const oneHourAgoEpoch = currentTimeEpoch - 3600000;
  const oneHourAgoISOString = new Date(oneHourAgoEpoch).toISOString();

  while (true) {
    const result = await fetch(
      `${NETHERLANDS_MEASUREMENTS_URL}?page=${measurementsPage}&formula=PM25&start=${oneHourAgoISOString}&end=${new Date().toISOString()}`
    );
    const resp: Response<Measurement[]> = await result.json();

    measurements.push(...resp.data);
    measurementsPage = resp.pagination.next_page;
    if (
      resp.pagination.current_page === resp.pagination.last_page ||
      resp.data.length === 0
    ) {
      break;
    }
  }

  const cities = measurements.map((measurement) => ({
    name:
      stations.find((stn) => stn.number === measurement.station_number)
        ?.location ?? '',
    data: Math.round(measurement.value),
  }));

  return Bluebird.map(
    cities,
    async (city) => {
      const location = await getLatLng(city.name);
      return { ...city, location };
    },
    { concurrency: 2 }
  );
};

File: internal/sources/netherlands.go
package sources

import (
	"context"
	"net/url"
	"strconv"
	"time"

	"github.com/undertideco/nebulo-scraper/internal/httpclient"
)

const (
	netherlandsStationsURL     = "https://api.luchtmeetnet.nl/open_api/stations"
	netherlandsMeasurementsURL = "https://api.luchtmeetnet.nl/open_api/measurements"
)

func scrapeNetherlands(ctx context.Context, r *Runner) ([]City, error) {
	stations, err := fetchNetherlandsPages[netherlandsStation](ctx, r, netherlandsStationsURL, nil)
	if err != nil {
		return nil, err
	}

	r.Logger.Printf("Netherlands: Retrieved %d stations", len(stations))

	stationLocations := make(map[string]string, len(stations))
	for _, station := range stations {
		stationLocations[station.Number] = station.Location
	}

	now := time.Now().UTC()
	values := url.Values{}
	values.Set("formula", "PM25")
	values.Set("start", now.Add(-time.Hour).Format(time.RFC3339Nano))
	values.Set("end", now.Format(time.RFC3339Nano))

	measurements, err := fetchNetherlandsPages[netherlandsMeasurement](ctx, r, netherlandsMeasurementsURL, values)
	if err != nil {
		return nil, err
	}

	geocoder, err := r.RequireGeocoder()
	if err != nil {
		return nil, err
	}

	return mapConcurrent(ctx, measurements, 2, func(ctx context.Context, measurement netherlandsMeasurement) (City, error) {
		name := stationLocations[measurement.StationNumber]
		location, err := geocoder.LatLng(ctx, name)
		if err != nil {
			return City{}, err
		}

		return City{
			Name:     name,
			Region:   "Netherlands",
			Data:     roundFloat(measurement.Value),
			Location: location,
		}, nil
	})
}

func fetchNetherlandsPages[T any](ctx context.Context, r *Runner, baseURL string, values url.Values) ([]T, error) {
	items := make([]T, 0)
	page := 1

	for {
		pageValues := url.Values{}
		for key, vals := range values {
			for _, val := range vals {
				pageValues.Add(key, val)
			}
		}
		pageValues.Set("page", strconv.Itoa(page))

		var resp netherlandsResponse[[]T]
		if err := httpclient.GetJSON(ctx, r.Client, baseURL+"?"+pageValues.Encode(), &resp); err != nil {
			return nil, err
		}

		items = append(items, resp.Data...)
		if resp.Pagination.CurrentPage == resp.Pagination.LastPage || len(resp.Data) == 0 {
			break
		}

		page = resp.Pagination.NextPage
	}

	return items, nil
}

type netherlandsStation struct {
	Number   string `json:"number"`
	Location string `json:"location"`
}

type netherlandsMeasurement struct {
	StationNumber     string  `json:"station_number"`
	Value             float64 `json:"value"`
	TimestampMeasured string  `json:"timestamp_measured"`
	Formula           string  `json:"formula"`
}

type netherlandsResponse[T any] struct {
	Pagination struct {
		LastPage    int `json:"last_page"`
		CurrentPage int `json:"current_page"`
		NextPage    int `json:"next_page"`
	} `json:"pagination"`
	Data T `json:"data"`
}
